python/pylibcudf/pylibcudf/tests/test_column_from_iterable.py

```diff
@@ -0,0 +1,168 @@
+    return request.param
+
+
+@pytest.fixture(params=DTYPES, ids=lambda x: x[0])
+def dtype_info(request):
+    return request.param
+
+
+def generate_list_data(shape, dtype):
+    def gen_values(n):
+        if dtype is bool:
+            return [(i % 2) == 0 for i in range(n)]
+        else:
+            return [dtype(i) for i in range(n)]
+
+    def nest(values, shape):
+        if not shape:
+            return values[0]
+        size = len(values) // shape[0]
+        return [
+            nest(values[i * size : (i + 1) * size], shape[1:])
+            for i in range(shape[0])
+        ]
+
+    return nest(gen_values(reduce(operator.mul, shape)), shape)
+
+
+def test_from_list(shape, dtype_info):
+    _, py_type, pa_type = dtype_info
+    data = generate_list_data(shape, py_type)
+    expect_type = pa_type
+    for _ in range(len(shape) - 1):
+        expect_type = pa.list_(expect_type)
+
+    expect = pa.array(data, type=expect_type)
+
+    got = plc.Column.from_iterable_of_py(
+        data, dtype=plc.interop.from_arrow(pa_type)
+    )
+
+    assert_column_eq(expect, got)
+
+
+def test_from_list_empty_without_dtype_raises():
+    with pytest.raises(
+        ValueError, match="Cannot infer dtype from empty iterable object"
+    ):
+        plc.Column.from_iterable_of_py([])
+
+
+def test_from_list_irregular_shapes_raises():
+    data = [[1], [2, 3], [], [4]]
+    with pytest.raises(ValueError, match="Inconsistent inner list shapes"):
+        plc.Column.from_iterable_of_py(data)
+
+
+def test_from_list_nested_dicts_raises():
+    data = [[{"a": 1}], [{"b": 2}]]
+    with pytest.raises(TypeError, match="Unsupported scalar type"):
+        plc.Column.from_iterable_of_py(data)
+
+
+def test_from_list_nested_strings_raises():
+    data = [["a", "b"], ["c", "d"]]
+    with pytest.raises(TypeError, match="Unsupported scalar type"):
+        plc.Column.from_iterable_of_py(data)
+
+
+def test_from_list_nested_none_raises():
+    data = [[None], [None]]
+    with pytest.raises(TypeError, match="Unsupported scalar type"):
+        plc.Column.from_iterable_of_py(data)
+
+
+def test_from_zero_dimensional_list_with_dtype():
+    got = plc.Column.from_iterable_of_py(
+        [], dtype=plc.DataType(plc.TypeId.STRING)
+    )
+    expect = pa.array([], type=pa.string())
+    assert_column_eq(expect, got)
+
+
+@pytest.mark.parametrize(
+    "factory, pa_type",
+    [
+        (lambda: range(5), pa.int64()),
+        (lambda: (x for x in range(5)), pa.float32()),
+        (lambda: array.array("q", range(5)), pa.int64()),
+        (lambda: filter(lambda x: x % 2 == 0, range(10)), pa.int16()),
+        (lambda: map(int, [1.1, 2.2, 3.3]), pa.int64()),
+        (lambda: reversed([1, 2, 3, 4]), pa.int64()),
+        (lambda: tuple(range(4)), pa.int32()),
+    ],
+)
+def test_from_builtin_iterables(factory, pa_type):
+    materialized = list(factory())
+    got = plc.Column.from_iterable_of_py(
+        iter(materialized), dtype=plc.interop.from_arrow(pa_type)
```

Comment:
Do we need to always call `iter` here? Is `array.array` not iterable?

Reply:
No I think we can remove it

Reply:
No I think we can remove it